{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Combine and Clean Retina Data\n",
    "\n",
    "The purpose of this notebook is to combine all the digital gene expression data for the retina cells, downloaded from the Gene Expression Omnibus using the accession number [GSE63473](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE63473)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Figure folder: ../figures/06_combine_retina_data\n",
      "Data folder: ../data/06_combine_retina_data\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import common\n",
    "\n",
    "# Assign notebook and folder names\n",
    "notebook_name = '06_combine_retina_data'\n",
    "figure_folder = os.path.join(common.FIGURE_FOLDER, notebook_name)\n",
    "data_folder = os.path.join(common.DATA_FOLDER, notebook_name)\n",
    "print('Figure folder:', figure_folder)\n",
    "print('Data folder:', data_folder)\n",
    "\n",
    "# Make the folders\n",
    "! mkdir -p $figure_folder\n",
    "! mkdir -p $data_folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'../data/00_original/GSM1626793_P14Retina_1.digital_expression.txt.gz'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_folder = os.path.join(common.DATA_FOLDER, '00_original')\n",
    "\n",
    "filename = os.path.join(input_folder, 'GSM1626793_P14Retina_1.digital_expression.txt.gz')\n",
    "filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSM1544798_SpeciesMix_ThousandSTAMPs_HUMAN.digital_expression.txt.gz\r\n",
      "GSM1544798_SpeciesMix_ThousandSTAMPs_MOUSE.digital_expression.txt.gz\r\n",
      "GSM1544799_SpeciesMix_HundredSTAMPs_HUMAN.digital_expression.txt.gz\r\n",
      "GSM1544799_SpeciesMix_HundredSTAMPs_MOUSE.digital_expression.txt.gz\r\n",
      "GSM1626793_P14Retina_1.digital_expression.txt.gz\r\n",
      "GSM1626794_P14Retina_2.digital_expression.txt.gz\r\n",
      "GSM1626795_P14Retina_3.digital_expression.txt.gz\r\n",
      "GSM1626796_P14Retina_4.digital_expression.txt.gz\r\n",
      "GSM1626797_P14Retina_5.digital_expression.txt.gz\r\n",
      "GSM1626798_P14Retina_6.digital_expression.txt.gz\r\n",
      "GSM1626799_P14Retina_7.digital_expression.txt.gz\r\n",
      "GSM1629192_Pure_HumanMouse_HUMAN.digital_expression.txt.gz\r\n",
      "GSM1629192_Pure_HumanMouse_MOUSE.digital_expression.txt.gz\r\n",
      "GSM1629193_ERCC.digital_expression.txt.gz\r\n",
      "GSM1629193_hg19_ERCC.dict.txt.gz\r\n",
      "GSM1629193_hg19_ERCC.refFlat.txt.gz\r\n",
      "mmc1.pdf\r\n",
      "mmc2.xlsx\r\n",
      "mmc3.xlsx\r\n",
      "mmc4.xlsx\r\n",
      "mmc4_v2.xlsx\r\n",
      "retina_clusteridentities.txt\r\n",
      "~$mmc4_v2.xlsx\r\n"
     ]
    }
   ],
   "source": [
    "ls $input_folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- gsm_i: 3, group_i: 1 ---\n",
      "\t ../data/00_original/GSM1626793_P14Retina_1.digital_expression.txt.gz\n",
      "CPU times: user 22.6 s, sys: 2.19 s, total: 24.8 s\n",
      "Wall time: 25 s\n",
      "--- gsm_i: 4, group_i: 2 ---\n",
      "\t ../data/00_original/GSM1626794_P14Retina_2.digital_expression.txt.gz\n",
      "CPU times: user 44.9 s, sys: 4.34 s, total: 49.2 s\n",
      "Wall time: 50 s\n",
      "--- gsm_i: 5, group_i: 3 ---\n",
      "\t ../data/00_original/GSM1626795_P14Retina_3.digital_expression.txt.gz\n",
      "CPU times: user 19.9 s, sys: 1.85 s, total: 21.7 s\n",
      "Wall time: 21.8 s\n",
      "--- gsm_i: 6, group_i: 4 ---\n",
      "\t ../data/00_original/GSM1626796_P14Retina_4.digital_expression.txt.gz\n",
      "CPU times: user 24.9 s, sys: 2.51 s, total: 27.4 s\n",
      "Wall time: 27.8 s\n",
      "--- gsm_i: 7, group_i: 5 ---\n",
      "\t ../data/00_original/GSM1626797_P14Retina_5.digital_expression.txt.gz\n",
      "CPU times: user 25 s, sys: 2.39 s, total: 27.4 s\n",
      "Wall time: 27.7 s\n",
      "--- gsm_i: 8, group_i: 6 ---\n",
      "\t ../data/00_original/GSM1626798_P14Retina_6.digital_expression.txt.gz\n",
      "CPU times: user 38 s, sys: 2.57 s, total: 40.5 s\n",
      "Wall time: 40.7 s\n",
      "--- gsm_i: 9, group_i: 7 ---\n",
      "\t ../data/00_original/GSM1626799_P14Retina_7.digital_expression.txt.gz\n",
      "CPU times: user 9.83 s, sys: 1 s, total: 10.8 s\n",
      "Wall time: 10.8 s\n",
      "expression.shape (49300, 24760)\n",
      "cell_metadata.shape (49300, 1)\n",
      "CPU times: user 3min 37s, sys: 28.9 s, total: 4min 6s\n",
      "Wall time: 4min 11s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "tables = []\n",
    "retina_numbers = zip(range(3, 10), range(1, 8))\n",
    "\n",
    "template = os.path.join(input_folder, 'GSM162679{}_P14Retina_{}.digital_expression.txt.gz')\n",
    "\n",
    "cell_metadata_dfs = []\n",
    "\n",
    "for gsm_i, group_i in retina_numbers:\n",
    "    print(f\"--- gsm_i: {gsm_i}, group_i: {group_i} ---\")\n",
    "    filename = template.format(gsm_i, group_i)\n",
    "    print('\\t', filename)\n",
    "    %time table = pd.read_table(filename, compression='gzip', index_col=0)\n",
    "    \n",
    "    # Transpose so genes are columns and cells are rows, creating a\n",
    "    # (samples, features) matrix\n",
    "    table = table.T\n",
    "    tables.append(table)\n",
    "    \n",
    "    df = pd.DataFrame(index=table.index)\n",
    "    df['batch'] = group_i\n",
    "    cell_metadata_dfs.append(df)\n",
    "    \n",
    "expression = pd.concat(tables)\n",
    "print('expression.shape', expression.shape)\n",
    "\n",
    "cell_metadata = pd.concat(cell_metadata_dfs)\n",
    "print('cell_metadata.shape', cell_metadata.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10:100015630-100100413:Kitl</th>\n",
       "      <th>10:100306571-100307185:Gm9476</th>\n",
       "      <th>10:100443902-100487350:Tmtc3</th>\n",
       "      <th>10:100488289-100573655:Cep290</th>\n",
       "      <th>10:100572274-100589259:4930430F08Rik</th>\n",
       "      <th>10:100592386-100618391:1700017N19Rik</th>\n",
       "      <th>10:10100953-10101116:Gm25682</th>\n",
       "      <th>10:101681487-102391469:Mgat4c</th>\n",
       "      <th>10:102481760-102490418:Nts</th>\n",
       "      <th>10:102512222-102546560:Rassf9</th>\n",
       "      <th>...</th>\n",
       "      <th>Y:1096861-1245759:Uty</th>\n",
       "      <th>Y:1260715-1286613:Ddx3y</th>\n",
       "      <th>Y:1298957-1459782:Usp9y</th>\n",
       "      <th>Y:45385648-45386331:Gm21779</th>\n",
       "      <th>Y:52674915-52675541:Gm21840</th>\n",
       "      <th>Y:78835860-78836543:Gm20806</th>\n",
       "      <th>Y:86365740-86366423:Gm20861</th>\n",
       "      <th>Y:897788-943811:Kdm5d</th>\n",
       "      <th>Y:90755657-90763485:Gm21857</th>\n",
       "      <th>Y:991630-991748:n-R5s1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GGCCGCAGTCCG</th>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTTGTGCGGGAA</th>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCGCAACTGCTC</th>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GATTGGGAGGCA</th>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCTCCTAGTTGG</th>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24760 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              10:100015630-100100413:Kitl  10:100306571-100307185:Gm9476  \\\n",
       "GGCCGCAGTCCG                            0                            NaN   \n",
       "CTTGTGCGGGAA                            0                            NaN   \n",
       "GCGCAACTGCTC                            1                            NaN   \n",
       "GATTGGGAGGCA                            0                            NaN   \n",
       "CCTCCTAGTTGG                            0                            NaN   \n",
       "\n",
       "              10:100443902-100487350:Tmtc3  10:100488289-100573655:Cep290  \\\n",
       "GGCCGCAGTCCG                             3                              1   \n",
       "CTTGTGCGGGAA                             0                              3   \n",
       "GCGCAACTGCTC                             0                              0   \n",
       "GATTGGGAGGCA                             0                              2   \n",
       "CCTCCTAGTTGG                             2                              1   \n",
       "\n",
       "              10:100572274-100589259:4930430F08Rik  \\\n",
       "GGCCGCAGTCCG                                     2   \n",
       "CTTGTGCGGGAA                                     1   \n",
       "GCGCAACTGCTC                                     2   \n",
       "GATTGGGAGGCA                                     0   \n",
       "CCTCCTAGTTGG                                     1   \n",
       "\n",
       "              10:100592386-100618391:1700017N19Rik  \\\n",
       "GGCCGCAGTCCG                                     0   \n",
       "CTTGTGCGGGAA                                     0   \n",
       "GCGCAACTGCTC                                     0   \n",
       "GATTGGGAGGCA                                     0   \n",
       "CCTCCTAGTTGG                                     0   \n",
       "\n",
       "              10:10100953-10101116:Gm25682  10:101681487-102391469:Mgat4c  \\\n",
       "GGCCGCAGTCCG                           NaN                              0   \n",
       "CTTGTGCGGGAA                           NaN                              0   \n",
       "GCGCAACTGCTC                           NaN                              4   \n",
       "GATTGGGAGGCA                           NaN                              1   \n",
       "CCTCCTAGTTGG                           NaN                              2   \n",
       "\n",
       "              10:102481760-102490418:Nts  10:102512222-102546560:Rassf9  \\\n",
       "GGCCGCAGTCCG                         NaN                              0   \n",
       "CTTGTGCGGGAA                         NaN                              0   \n",
       "GCGCAACTGCTC                         NaN                              0   \n",
       "GATTGGGAGGCA                         NaN                              0   \n",
       "CCTCCTAGTTGG                         NaN                              0   \n",
       "\n",
       "                       ...            Y:1096861-1245759:Uty  \\\n",
       "GGCCGCAGTCCG           ...                                0   \n",
       "CTTGTGCGGGAA           ...                                1   \n",
       "GCGCAACTGCTC           ...                                0   \n",
       "GATTGGGAGGCA           ...                                1   \n",
       "CCTCCTAGTTGG           ...                                0   \n",
       "\n",
       "              Y:1260715-1286613:Ddx3y  Y:1298957-1459782:Usp9y  \\\n",
       "GGCCGCAGTCCG                        0                      NaN   \n",
       "CTTGTGCGGGAA                        5                      NaN   \n",
       "GCGCAACTGCTC                        0                      NaN   \n",
       "GATTGGGAGGCA                        0                      NaN   \n",
       "CCTCCTAGTTGG                        0                      NaN   \n",
       "\n",
       "              Y:45385648-45386331:Gm21779  Y:52674915-52675541:Gm21840  \\\n",
       "GGCCGCAGTCCG                          NaN                          NaN   \n",
       "CTTGTGCGGGAA                          NaN                          NaN   \n",
       "GCGCAACTGCTC                          NaN                          NaN   \n",
       "GATTGGGAGGCA                          NaN                          NaN   \n",
       "CCTCCTAGTTGG                          NaN                          NaN   \n",
       "\n",
       "              Y:78835860-78836543:Gm20806  Y:86365740-86366423:Gm20861  \\\n",
       "GGCCGCAGTCCG                          NaN                          NaN   \n",
       "CTTGTGCGGGAA                          NaN                          NaN   \n",
       "GCGCAACTGCTC                          NaN                          NaN   \n",
       "GATTGGGAGGCA                          NaN                          NaN   \n",
       "CCTCCTAGTTGG                          NaN                          NaN   \n",
       "\n",
       "              Y:897788-943811:Kdm5d  Y:90755657-90763485:Gm21857  \\\n",
       "GGCCGCAGTCCG                      0                          NaN   \n",
       "CTTGTGCGGGAA                      0                          NaN   \n",
       "GCGCAACTGCTC                      0                          NaN   \n",
       "GATTGGGAGGCA                      0                          NaN   \n",
       "CCTCCTAGTTGG                      0                          NaN   \n",
       "\n",
       "              Y:991630-991748:n-R5s1  \n",
       "GGCCGCAGTCCG                     0.0  \n",
       "CTTGTGCGGGAA                     0.0  \n",
       "GCGCAACTGCTC                     0.0  \n",
       "GATTGGGAGGCA                     0.0  \n",
       "CCTCCTAGTTGG                     0.0  \n",
       "\n",
       "[5 rows x 24760 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GGCCGCAGTCCG</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTTGTGCGGGAA</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCGCAACTGCTC</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GATTGGGAGGCA</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCTCCTAGTTGG</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              batch\n",
       "GGCCGCAGTCCG      1\n",
       "CTTGTGCGGGAA      1\n",
       "GCGCAACTGCTC      1\n",
       "GATTGGGAGGCA      1\n",
       "CCTCCTAGTTGG      1"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "batch\n",
       "1    6600\n",
       "2    9000\n",
       "3    6120\n",
       "4    7650\n",
       "5    7650\n",
       "6    8280\n",
       "7    4000\n",
       "dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_metadata.groupby('batch').size()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Clean expression matrix to be compatible with the cluster labels and identities\n",
    "\n",
    "Currently, cells are labeled by their barcode, e.g. `GCGCAACTGCTC`, and genes are labeled by their chrom:start-end:symbol, e.g. `6:51460434-51469894:Hnrnpa2b1`. But, in the supplementary data, the genes are all uppercase, e.g. `HNRNPA2B1` (which is incorrect since this is mouse data.. ) and the barcodes have `r1_` prepended before the id, e.g. `r1_GCGCAACTGCTC`.\n",
    "\n",
    "So we need to clean the data to be compatible with this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>symbol</th>\n",
       "      <th>KITL</th>\n",
       "      <th>GM9476</th>\n",
       "      <th>TMTC3</th>\n",
       "      <th>CEP290</th>\n",
       "      <th>4930430F08RIK</th>\n",
       "      <th>1700017N19RIK</th>\n",
       "      <th>GM25682</th>\n",
       "      <th>MGAT4C</th>\n",
       "      <th>NTS</th>\n",
       "      <th>RASSF9</th>\n",
       "      <th>...</th>\n",
       "      <th>UTY</th>\n",
       "      <th>DDX3Y</th>\n",
       "      <th>USP9Y</th>\n",
       "      <th>GM21779</th>\n",
       "      <th>GM21840</th>\n",
       "      <th>GM20806</th>\n",
       "      <th>GM20861</th>\n",
       "      <th>KDM5D</th>\n",
       "      <th>GM21857</th>\n",
       "      <th>N-R5S1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GGCCGCAGTCCG</th>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTTGTGCGGGAA</th>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCGCAACTGCTC</th>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GATTGGGAGGCA</th>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCTCCTAGTTGG</th>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24760 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "symbol        KITL  GM9476  TMTC3  CEP290  4930430F08RIK  1700017N19RIK  \\\n",
       "GGCCGCAGTCCG     0     NaN      3       1              2              0   \n",
       "CTTGTGCGGGAA     0     NaN      0       3              1              0   \n",
       "GCGCAACTGCTC     1     NaN      0       0              2              0   \n",
       "GATTGGGAGGCA     0     NaN      0       2              0              0   \n",
       "CCTCCTAGTTGG     0     NaN      2       1              1              0   \n",
       "\n",
       "symbol        GM25682  MGAT4C  NTS  RASSF9   ...    UTY  DDX3Y  USP9Y  \\\n",
       "GGCCGCAGTCCG      NaN       0  NaN       0   ...      0      0    NaN   \n",
       "CTTGTGCGGGAA      NaN       0  NaN       0   ...      1      5    NaN   \n",
       "GCGCAACTGCTC      NaN       4  NaN       0   ...      0      0    NaN   \n",
       "GATTGGGAGGCA      NaN       1  NaN       0   ...      1      0    NaN   \n",
       "CCTCCTAGTTGG      NaN       2  NaN       0   ...      0      0    NaN   \n",
       "\n",
       "symbol        GM21779  GM21840  GM20806  GM20861  KDM5D  GM21857  N-R5S1  \n",
       "GGCCGCAGTCCG      NaN      NaN      NaN      NaN      0      NaN     0.0  \n",
       "CTTGTGCGGGAA      NaN      NaN      NaN      NaN      0      NaN     0.0  \n",
       "GCGCAACTGCTC      NaN      NaN      NaN      NaN      0      NaN     0.0  \n",
       "GATTGGGAGGCA      NaN      NaN      NaN      NaN      0      NaN     0.0  \n",
       "CCTCCTAGTTGG      NaN      NaN      NaN      NaN      0      NaN     0.0  \n",
       "\n",
       "[5 rows x 24760 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_symbols = expression.columns.map(lambda x: x.split(':')[-1].upper())\n",
    "gene_symbols.name = 'symbol'\n",
    "expression.columns = gene_symbols\n",
    "expression.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# %%time\n",
    "# csv = os.path.join(data_folder, 'retina_expression.csv')\n",
    "# expression.to_csv(csv)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "CPU times: user 59min 38s, sys: 1min 53s, total: 1h 1min 31s\n",
    "Wall time: 2h 31min 41s\n",
    "```\n",
    "\n",
    "## 2h 31m to write a csv file -- Woww!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import xarray as xr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(49300, 24760)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['KITL', 'GM9476', 'TMTC3', 'CEP290', '4930430F08RIK', '1700017N19RIK',\n",
       "       'GM25682', 'MGAT4C', 'NTS', 'RASSF9'],\n",
       "      dtype='object', name='symbol')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression.columns[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add retinal cell cluster metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_01</th>\n",
       "      <th>cluster_02</th>\n",
       "      <th>cluster_03</th>\n",
       "      <th>cluster_04</th>\n",
       "      <th>cluster_05</th>\n",
       "      <th>cluster_06</th>\n",
       "      <th>cluster_07</th>\n",
       "      <th>cluster_08</th>\n",
       "      <th>cluster_09</th>\n",
       "      <th>cluster_10</th>\n",
       "      <th>...</th>\n",
       "      <th>cluster_30</th>\n",
       "      <th>cluster_31</th>\n",
       "      <th>cluster_32</th>\n",
       "      <th>cluster_33</th>\n",
       "      <th>cluster_34</th>\n",
       "      <th>cluster_35</th>\n",
       "      <th>cluster_36</th>\n",
       "      <th>cluster_37</th>\n",
       "      <th>cluster_38</th>\n",
       "      <th>cluster_39</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>r1_GGCCGCAGTCCG</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>r1_CTTGTGCGGGAA</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>r1_GCGCAACTGCTC</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>r1_GATTGGGAGGCA</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>r1_GTGCCGCCTCTC</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 cluster_01  cluster_02  cluster_03  cluster_04  cluster_05  \\\n",
       "cell                                                                          \n",
       "r1_GGCCGCAGTCCG           0           1           0           0           0   \n",
       "r1_CTTGTGCGGGAA           0           1           0           0           0   \n",
       "r1_GCGCAACTGCTC           0           1           0           0           0   \n",
       "r1_GATTGGGAGGCA           0           1           0           0           0   \n",
       "r1_GTGCCGCCTCTC           0           0           0           0           0   \n",
       "\n",
       "                 cluster_06  cluster_07  cluster_08  cluster_09  cluster_10  \\\n",
       "cell                                                                          \n",
       "r1_GGCCGCAGTCCG           0           0           0           0           0   \n",
       "r1_CTTGTGCGGGAA           0           0           0           0           0   \n",
       "r1_GCGCAACTGCTC           0           0           0           0           0   \n",
       "r1_GATTGGGAGGCA           0           0           0           0           0   \n",
       "r1_GTGCCGCCTCTC           0           0           0           0           0   \n",
       "\n",
       "                    ...      cluster_30  cluster_31  cluster_32  cluster_33  \\\n",
       "cell                ...                                                       \n",
       "r1_GGCCGCAGTCCG     ...               0           0           0           0   \n",
       "r1_CTTGTGCGGGAA     ...               0           0           0           0   \n",
       "r1_GCGCAACTGCTC     ...               0           0           0           0   \n",
       "r1_GATTGGGAGGCA     ...               0           0           0           0   \n",
       "r1_GTGCCGCCTCTC     ...               0           0           0           0   \n",
       "\n",
       "                 cluster_34  cluster_35  cluster_36  cluster_37  cluster_38  \\\n",
       "cell                                                                          \n",
       "r1_GGCCGCAGTCCG           0           0           0           0           0   \n",
       "r1_CTTGTGCGGGAA           0           0           0           0           0   \n",
       "r1_GCGCAACTGCTC           0           0           0           0           0   \n",
       "r1_GATTGGGAGGCA           0           0           0           0           0   \n",
       "r1_GTGCCGCCTCTC           0           0           0           0           0   \n",
       "\n",
       "                 cluster_39  \n",
       "cell                         \n",
       "r1_GGCCGCAGTCCG           0  \n",
       "r1_CTTGTGCGGGAA           0  \n",
       "r1_GCGCAACTGCTC           0  \n",
       "r1_GATTGGGAGGCA           0  \n",
       "r1_GTGCCGCCTCTC           0  \n",
       "\n",
       "[5 rows x 39 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "csv = os.path.join(common.DATA_FOLDER, '03_clean_cluster_assignments', \n",
    "                   'cluster_bools.csv')\n",
    "cluster_bools = pd.read_csv(csv, index_col=0, squeeze=True)\n",
    "cluster_bools.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cell\n",
       "r1_GGCCGCAGTCCG    cluster_02\n",
       "r1_CTTGTGCGGGAA    cluster_02\n",
       "r1_GCGCAACTGCTC    cluster_02\n",
       "r1_GATTGGGAGGCA    cluster_02\n",
       "r1_GTGCCGCCTCTC    cluster_25\n",
       "Name: cluster_id, dtype: object"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "csv = os.path.join(common.DATA_FOLDER, '03_clean_cluster_assignments', \n",
    "                   'cluster_names.csv')\n",
    "cluster_names = pd.read_csv(csv, index_col=0, squeeze=True)\n",
    "cluster_names.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GGCCGCAGTCCG</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTTGTGCGGGAA</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCGCAACTGCTC</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GATTGGGAGGCA</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCTCCTAGTTGG</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              batch\n",
       "GGCCGCAGTCCG      1\n",
       "CTTGTGCGGGAA      1\n",
       "GCGCAACTGCTC      1\n",
       "GATTGGGAGGCA      1\n",
       "CCTCCTAGTTGG      1"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "batch    int64\n",
       "dtype: object"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_metadata.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(49300, 1)\n",
      "(0, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>cluster_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [batch, cluster_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(cell_metadata.shape)\n",
    "cell_metadata_clusters = cell_metadata.join(cluster_names, how='inner')\n",
    "print(cell_metadata_clusters.shape)\n",
    "cell_metadata_clusters.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Can't convert to int because some are NA\n",
    "# cell_metadata_clusters['cluster_id'] = cell_metadata_clusters['cluster_id'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "batch          int64\n",
       "cluster_id    object\n",
       "dtype: object"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_metadata_clusters.dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add gene metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>retina_01</th>\n",
       "      <th>retina_02</th>\n",
       "      <th>retina_03</th>\n",
       "      <th>retina_04</th>\n",
       "      <th>retina_05</th>\n",
       "      <th>retina_06</th>\n",
       "      <th>retina_07</th>\n",
       "      <th>retina_08</th>\n",
       "      <th>retina_09</th>\n",
       "      <th>retina_10</th>\n",
       "      <th>...</th>\n",
       "      <th>retina_38</th>\n",
       "      <th>retina_39</th>\n",
       "      <th>cellcycle_01</th>\n",
       "      <th>cellcycle_02</th>\n",
       "      <th>cellcycle_03</th>\n",
       "      <th>cellcycle_04</th>\n",
       "      <th>cellcycle_05</th>\n",
       "      <th>cellcycle_06</th>\n",
       "      <th>cellcycle_07</th>\n",
       "      <th>cellcycle_08</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Aaas</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Acat2</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Acot9</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Actb</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Adar</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 47 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             retina_01  retina_02  retina_03  retina_04  retina_05  retina_06  \\\n",
       "gene_symbol                                                                     \n",
       "    Aaas         False      False      False      False      False      False   \n",
       "    Acat2        False      False      False      False      False      False   \n",
       "    Acot9        False      False      False      False      False      False   \n",
       "    Actb         False      False      False      False      False      False   \n",
       "    Adar         False      False      False      False      False      False   \n",
       "\n",
       "             retina_07  retina_08  retina_09  retina_10      ...       \\\n",
       "gene_symbol                                                  ...        \n",
       "    Aaas         False      False      False      False      ...        \n",
       "    Acat2        False      False      False      False      ...        \n",
       "    Acot9        False      False      False      False      ...        \n",
       "    Actb         False      False      False      False      ...        \n",
       "    Adar         False      False      False      False      ...        \n",
       "\n",
       "             retina_38  retina_39  cellcycle_01  cellcycle_02  cellcycle_03  \\\n",
       "gene_symbol                                                                   \n",
       "    Aaas         False      False         False         False         False   \n",
       "    Acat2        False      False         False         False          True   \n",
       "    Acot9        False      False         False         False         False   \n",
       "    Actb         False      False         False         False         False   \n",
       "    Adar         False      False         False         False         False   \n",
       "\n",
       "             cellcycle_04  cellcycle_05  cellcycle_06  cellcycle_07  \\\n",
       "gene_symbol                                                           \n",
       "    Aaas            False          True         False         False   \n",
       "    Acat2           False         False         False         False   \n",
       "    Acot9           False         False         False          True   \n",
       "    Actb            False         False         False         False   \n",
       "    Adar            False         False         False         False   \n",
       "\n",
       "             cellcycle_08  \n",
       "gene_symbol                \n",
       "    Aaas            False  \n",
       "    Acat2           False  \n",
       "    Acot9           False  \n",
       "    Actb             True  \n",
       "    Adar             True  \n",
       "\n",
       "[5 rows x 47 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "csv = os.path.join(common.DATA_FOLDER, \n",
    "                   '04_extract_data_from_supplementary_excel_files', \n",
    "                   'mouse_gene_metadata.csv')\n",
    "gene_metadata = pd.read_csv(csv, index_col=0)\n",
    "gene_metadata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make subsets for teaching\n",
    "\n",
    "### Dropout demo - Equal sized clusters\n",
    "\n",
    "only use cells from the first batch, and only use differentially expressed genes from the biggest clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'cluster_id'",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m~/anaconda3/envs/macosko2015-env/lib/python3.6/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_value\u001b[0;34m(self, series, key)\u001b[0m\n\u001b[1;32m   2482\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2483\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mlibts\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_value_box\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2484\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mIndexError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/tslib.pyx\u001b[0m in \u001b[0;36mpandas._libs.tslib.get_value_box (pandas/_libs/tslib.c:18843)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/tslib.pyx\u001b[0m in \u001b[0;36mpandas._libs.tslib.get_value_box (pandas/_libs/tslib.c:18477)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: 'str' object cannot be interpreted as an integer",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-22-288ac7991c12>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mcluster_sizes_table0\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtables\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcluster_names\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'cluster_id'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mcluster_sizes_table0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/macosko2015-env/lib/python3.6/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    599\u001b[0m         \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_apply_if_callable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    600\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 601\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_value\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    602\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    603\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mis_scalar\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/macosko2015-env/lib/python3.6/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_value\u001b[0;34m(self, series, key)\u001b[0m\n\u001b[1;32m   2489\u001b[0m                     \u001b[0;32mraise\u001b[0m \u001b[0mInvalidIndexError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2490\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2491\u001b[0;31m                     \u001b[0;32mraise\u001b[0m \u001b[0me1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2492\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# pragma: no cover\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2493\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0me1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/macosko2015-env/lib/python3.6/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_value\u001b[0;34m(self, series, key)\u001b[0m\n\u001b[1;32m   2475\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2476\u001b[0m             return self._engine.get_value(s, k,\n\u001b[0;32m-> 2477\u001b[0;31m                                           tz=getattr(series.dtype, 'tz', None))\n\u001b[0m\u001b[1;32m   2478\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2479\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minferred_type\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'integer'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'boolean'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_value (pandas/_libs/index.c:4404)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_value (pandas/_libs/index.c:4087)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc (pandas/_libs/index.c:5126)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item (pandas/_libs/hashtable.c:20523)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item (pandas/_libs/hashtable.c:20477)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'cluster_id'"
     ],
     "output_type": "error"
    }
   ],
   "source": [
    "cluster_sizes_table0 = tables[0].groupby(cluster_names['cluster_id'], axis=0).size()\n",
    "cluster_sizes_table0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "big_clusters = cluster_sizes_table0[cluster_sizes_table0 > 100]\n",
    "big_clusters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make superset of all genes for posterity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Need to align the cell metadata with expression so xarray doesn't get mad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(cell_metadata_clusters.shape)\n",
    "\n",
    "cells_left, expression_right = cell_metadata_clusters.align(\n",
    "    expression, axis=0, join='inner')\n",
    "print(cells_left.shape, expression_right.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Align the gene metadata too"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genes_t = gene_metadata.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_t_left, expression_genes_right = genes_t.align(expression_right, axis=1, join='right')\n",
    "print(genes_t_left.shape, expression_genes_right.shape)\n",
    "genes_t_left.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genes_left = genes_t_left.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cells_left.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ds = xr.Dataset(\n",
    "#     {'expression': (['cells', 'genes'], expression_genes_right),\n",
    "#      'cell_metadata': (['cells', 'cell_features', ], cells_left),\n",
    "#      'gene_metadata': (['genes', 'gene_features', ], genes_left),\n",
    "#     })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "range(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "netcdf = os.path.join(data_folder, 'retina_all_genes.netcdf')\n",
    "ds.to_netcdf(netcdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ls -lha $netcdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "netcdf = os.path.join(data_folder, 'retina.netcdf')\n",
    "ds.to_netcdf(netcdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ls -lha $netcdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    ""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (macosko2015-env)",
   "language": "python",
   "name": "macosko2015-env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3.0
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}